### ISSUE TITLE
ARM64e iOS/macOS pointer authentication

### ISSUE THREAD

**Issue by houweifeng** [LuaJIT/LuaJIT#559]
Incident Identifier: 694BB9CB-4581-4510-A8F4-2E9649B69355
CrashReporter Key:   40896ec6605b2bd558fe1688e28ff24afb392bb1
Hardware Model:      iPhone11,8
Process:             luajit-arm64e [4830]
Path:                /private/var/root/houwf/tools/luajit/luajit-arm64e
Identifier:          luajit-arm64e
Version:             ???
Code Type:           ARM-64 (Native)
Role:                Unspecified
Parent Process:      sh [4803]
Coalition:           com.openssh.sshd.75E05888-38EF-4492-9ED4-EEB579525CCC [1426]

Date/Time:           2020-02-26 23:26:55.0709 +0800
Launch Time:         2020-02-26 23:26:54.8249 +0800
OS Version:          iPhone OS 13.3 (17C54)
Release Type:        User
Baseband Version:    2.03.07
Report Version:      104

Exception Type:  EXC_BAD_ACCESS (SIGSEGV)
Exception Subtype: KERN_INVALID_ADDRESS at 0xa21e0e81007ad778 -> 0x00000001007ad778 (possible pointer authentication failure)
VM Region Info: 0x1007ad778 is in 0x10079c000-0x1007e4000;  bytes after start: 71544  bytes before end: 223367
      REGION TYPE                      START - END             [ VSIZE] PRT/MAX SHRMOD  REGION DETAIL
      UNUSED SPACE AT START
--->  __TEXT                 000000010079c000-00000001007e4000 [  288K] r-x/r-x SM=COW  ...luajit-arm64e
      __DATA                 00000001007e4000-00000001007e8000 [   16K] rw-/rw- SM=COW  ...luajit-arm64e

Triggered by Thread:  0

Thread 0 name:  Dispatch queue: com.apple.main-thread
Thread 0 Crashed:
0   ???                           	0xffffff81007ad778 0 + -545452796040
1   luajit-arm64e                 	0x00000001007ad748 0x10079c000 + 71496
2   luajit-arm64e                 	0x00000001007d3e10 0x10079c000 + 228880
3   luajit-arm64e                 	0x00000001007a0180 0x10079c000 + 16768
4   libdyld.dylib                 	0x0000000192b58e18 start + 4

========
https://developer.apple.com/documentation/security/preparing_your_app_to_work_with_pointer_authentication?language=objc

**Comment by MikePall** [comment]
If anybody wants to work on this: it's doable, but hard. Try to find all the spots where the pointer auth/deauth needs to be done inside the interpreter.

**Comment by iryont** [comment]
This issue will likely become more relevant when Apple releases its ARM based MacBook.

**Comment by houweifeng** [comment]
more information:
========
`lj_vm_cpcall:
    0x104780b6c <+0>:   stp    x29, x30, [sp, #-0xd0]!
    0x104780b70 <+4>:   mov    x29, sp
    0x104780b74 <+8>:   stp    x19, x20, [sp, #0x10]
    0x104780b78 <+12>:  stp    d8, d9, [sp, #0x60]
    0x104780b7c <+16>:  stp    x21, x22, [sp, #0x20]
    0x104780b80 <+20>:  stp    d10, d11, [sp, #0x70]
    0x104780b84 <+24>:  stp    x23, x24, [sp, #0x30]
    0x104780b88 <+28>:  stp    d12, d13, [sp, #0x80]
    0x104780b8c <+32>:  stp    x25, x26, [sp, #0x40]
    0x104780b90 <+36>:  stp    d14, d15, [sp, #0x90]
    0x104780b94 <+40>:  stp    x27, x28, [sp, #0x50]
    0x104780b98 <+44>:  mov    x23, x0
    0x104780b9c <+48>:  ldr    x27, [x0, #0x38]
    0x104780ba0 <+52>:  str    x0, [sp, #0xb0]
    0x104780ba4 <+56>:  ldr    x22, [x23, #0x10]
    0x104780ba8 <+60>:  ldr    x17, [x23, #0x28]
    0x104780bac <+64>:  str    x0, [sp, #0xa8]
    0x104780bb0 <+68>:  ldr    x28, [x23, #0x50]
    0x104780bb4 <+72>:  sub    x27, x27, x17
    0x104780bb8 <+76>:  str    w27, [sp, #0xc8]
    0x104780bbc <+80>:  str    wzr, [sp, #0xc4]
    0x104780bc0 <+84>:  str    x28, [sp, #0xa0]
    0x104780bc4 <+88>:  str    x29, [x23, #0x50]
    0x104780bc8 <+92>:  str    x23, [x22, #0x170]
    0x104780bcc <+96>:  blr    x3
->  0x104780bd0 <+100>: mov    x19, x0   <===== crash
=======
vm_arm64.dasc:

  |->vm_cpcall:				// Setup protected C frame, call C.
  |  // (lua_State *L, lua_CFunction func, void *ud, lua_CPFunction cp)
  |  saveregs
  |  mov L, CARG1
  |   ldr RA, L:CARG1->stack
  |  str CARG1, SAVE_L
  |    ldr GL, L->glref			// Setup pointer to global state.
  |   ldr RB, L->top
  |  str CARG1, SAVE_PC			// Any value outside of bytecode is ok.
  |  ldr RC, L->cframe
  |   sub RA, RA, RB			// Compute -savestack(L, L->top).
  |   str RAw, SAVE_NRES		// Neg. delta means cframe w/o frame.
  |  str wzr, SAVE_ERRF			// No error function.
  |  str RC, SAVE_CFRAME
  |  str fp, L->cframe			// Add our C frame to cframe chain.
  |    str L, GL->cur_L
  |  blr CARG4			// (lua_State *L, lua_CFunction func, void *ud)
  |  mov BASE, CRET1  <===== crash
======
It's just assembly code, isn't it? Other code does not need to be modified.

https://events.static.linuxfound.org/sites/events/files/slides/slides_23.pdf

**Comment by MikePall** [comment]
You're vastly underestimating what's needed to complete this task. As I said above, this is **hard**.

If anybody has the required skills and wants to contribute that code then report back here when you're done, not before.

**Comment by corsix** [comment]
> This issue will likely become more relevant when Apple releases its ARM based MacBook.

At the moment, the M1/M2 MacBooks have hardware support for ARM64e, but it is not enabled by default, and [it is deliberately difficult to enable](https://github.com/lelegard/arm-cpusysregs/blob/main/docs/arm64e-on-macos.md#enabling-arm64e-on-macos-13-ventura). That said, it is more interesting to work against the MacBook implementation than the iPhone implementation, as the former has JIT support.

> It's just assembly code, isn't it? Other code does not need to be modified.

In practice, a few other places need to be modified:
1. The assembler itself needs support for the relevant PAC instructions.
2. The error unwinder calls `_Unwind_SetIP`, which seemingly faults if passed a signed pointer, so the signing (as done by Clang on all function pointers) needs to be stripped before calling.
3. If JIT support is desired, then a number of things need changing:
    1. Function pointers and data pointers cannot be so easily interchanged, as the former will be signed by default. This cuts both ways; some places which use `void*` (generic pointer for dereferencing and/or arithmetic, no calling) should instead use `ASMFunction` (generic pointer for calling, no dereferencing and/or arithmetic), whereas the converse needs doing to `lj_vm_exit_handler`/`lj_vm_exit_interp`, as they're used for arithmetic rather than calling.
    2. The JIT assembler needs support for the relevant PAC instructions, and the JIT should use them.
    3. Yet more unwinding fun; the function pointer within the mcode-area CIE needs special signing (otherwise `__register_frame` faults, which is fun).

I've put something together at https://github.com/corsix/LuaJIT/commits/arm64e, though my interest is mostly academic - I have no strong need or urge for this to be merged upstream.

Inserting `pacibsp` instructions (and associated `ret` -> `retab` transformation) is possibly not strictly required, though I can imagine that future unwinders might be confused by frame records containing unsigned pointers. Potentially more contentious is signing the pointers in the interpreter's dispatch table - I've opted for doing so, even though it has performance implications on any chip where `braaz` is more expensive than `br`.

**Comment by MikePall** [comment]
Looks great, Peter!

However, how do we proceed? I don't have (nor want) the hardware to test this on a regular basis. AFAIK, there's no QEMU with ARM64e support (yet). And Peter, as stated, doesn't have an interest (nor the time) to maintain it.

Maintaining such a feature has a cost. There are likely some more odd places that need fixups. And there will be future changes to the VM that need to be carefully evaluated and tested with this feature in mind.

It would be a shame if Peter's work would bitrot due to not merging it now. OTOH someone who has a vested interest needs to bear the cost. I.e. continuous testing and manual supervision of all commits. The latter being the more costly part. Any takers?

**Comment by corsix** [comment]
To the best of my knowledge (which may well be incorrect or incomplete), two things are true at the moment for M1 macOS:
1. arm64e _applications_ can only run if they're first-party (which Tarantool isn't) or if system integrity is disabled and bootargs are changed appropriately (which seems like an unusual configuration). [NB: I've seen rumours that some M1 hardware in China is configured to let third-party arm64e _applications_ run by default, but I have no take on the truthiness of this]
2. arm64e _libraries_ can exist, but if loaded from an arm64 application, pointer signing instructions become the identity function, pointer authentication instructions always succeed, and combined instructions behave like their unauthenticated equivalent (`braaz` behaves like `br`, `retab` behaves like `ret`, etc.) [NB: Possibly a weaker form of this is - or will become - true, where instructions using the "A" keys are neutered as described, but instructions using the "B" keys are not neutered, meaning that return addresses are signed but general function pointers are not]

These points conspire to mean that a LuaJIT _library_ could be compiled as arm64e, but you wouldn't end up testing the arm64e-ness of it, as issues can only appear when pointer signing instructions actually do something.

With those points in mind, are you able to elaborate on why this is interesting for Tarantool at the current time?

**Comment by igormunkin** [comment]
Peter, I can hardly explain our interest in other way than Mike’s already done: we can’t expect what’s going to be done by Cupertino guys regarding this feature, but I also expect enabling this as a default target in a couple of years.

We recently focused on ARM64 support partly due to M1 release and we were in bit a rush for us as for all others lacking full ABI spec beforehand. All in all, there are still some things unresolved for JIT (hence it’s disabled by default) and it’s now a fully supported target at the moment.

We faced [an issue with Apple's libunwind](https://github.com/tarantool/tarantool/issues/8074) this year, and though [the fix is quite straightforward](https://github.com/tarantool/tarantool/pull/8818), it required resources to investigate the root cause. We use some other low level libraries in Tarantool and I prefer to avoid fixing their integration layers when arm64e target becomes more widely used.

Most of the Tarantool sources (i.e. everything except small hunks with handcrafted assembly and LuaJIT) can be simply rebuilt against the new target, hence LuaJIT is the major blocker for us.

**TL;DR: we just want to be in sync with outer world rather expecting benefits from PAC.** However, it still looks worthing to test Tarantool with PAC enabled to be sure we do nothing illegal.

Hence, from my side I’m for moving forward with your changes to be ready for any Apple decision to be made in the future.

I also would like to explicitly mention (considering Mike's comment above) the fact that we're interested only in M1 support (though I believe, this part is common for both iOS and macOS). And once again (if we're considered as the candidates for supervising this part): I want to know what exactly is expected from us regarding the patches.

**Comment by MikePall** [comment]
I've added this now. Thanks Peter!

@igormunkin My request for your team would be to regularly check if any future commits I make accidentally break ARM64e support. In this case, please open an issue here. Thanks!

**Comment by igormunkin** [comment]
Mike, thanks for applying this!

We'll do the regular testing for v2.1 branch in our CI: I'll share the dedicated branch/workflow here, when we're all set.

Unfortunately, I've tried to build the latest commit with the "hardening" support on my M1 laptop and the compilation error occurred (one can find the steps reproducing the issue below). 
```
$ git log -1 # the latest revision                                                        
commit 72efc42ef2258086a9cb797c676e2916b0a9e7e1
Author: Mike Pall <mike>
Date:   Sun Aug 13 03:02:25 2023 +0200

    MIPS: Fix "bad FP FLOAD" assertion.
    
    Reported by Sergey Kaplun. #1043
$ git status # no additional changes applied                                                                 
On branch v2.1
Your branch is up to date with 'origin/v2.1'.

nothing to commit, working tree clean
$ export MACOSX_DEPLOYMENT_TARGET=$(sw_vers -productVersion)                         
$ CFLAGS="-target arm64e-apple-macos" LDFLAGS="-target arm64e-apple-macos" make      
==== Building LuaJIT 2.1.0-beta3 ====
/Library/Developer/CommandLineTools/usr/bin/make -C src
HOSTCC    host/minilua.o
HOSTLINK  host/minilua
DYNASM    host/buildvm_arch.h
make[1]: *** [host/buildvm_arch.h] Segmentation fault: 11
make: *** [default] Error 2
```

@corsix, I would be thankful, if you share the right spell, you were using for testing your patch (I found nothing except [this answer](https://developer.apple.com/forums/thread/650248) on the Apple Developer Forum). Otherwise, we have a problem here D:

**Comment by igormunkin** [comment]
PEBKAC. It occurs that `arm64e` applications are not allowed to run on an `arm64` system. Many details can be found [here](https://github.com/lelegard/arm-cpusysregs/blob/main/docs/arm64e-on-macos.md#trying-the-arm64e-target-platform) (thanks @Gumix for sharing the docs). The recipe for enabling `arm64e` is also described within the aforementioned doc.

**Comment by MikePall** [comment]
Technically, you're cross-compiling from an ARM64 host to an ARM64e target. Because the build process creates `minilua` and other host binaries that need to be run on the host, which is not ARM64e.
I guess you'll need to set `TARGET_FLAGS`, which is added to both compiler and linker target flags.
